src/mscorlib/shared/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -686,6 +686,115 @@ protected EventSource(EventSourceSettings settings, params string[] traits)
+                byte[] eventNameBytes = Encoding.ASCII.GetBytes(eventName);
+
+                // evnetID          : 4 bytes
+                // eventName        : eventNameBytes.Length + 1
+                // keywords         : 8 bytes
+                // eventVersion     : 4 bytes
+                // level            : 4 bytes
+                uint metadataLength = 21 + (uint)eventNameBytes.Length;
+
+                // Increase the metadataLength for the types of all parameters.
+                metadataLength += (uint)m_eventData[i].Parameters.Length * 4;
+
+                // Increase the metadataLength for the names of all parameters.
+                foreach (var parameter in m_eventData[i].Parameters)
+                {
+                    string parameterName = parameter.Name;
+                    byte[] parameterNameBytes = Encoding.ASCII.GetBytes(parameterName);
+                    metadataLength = metadataLength + (uint)parameterNameBytes.Length + 1;
+                }
+
+                byte[] metadata = new byte[metadataLength];
+
+                // Write metadata: evnetID, eventName, keywords, eventVersion, level, param1 type, param1 name...
+                fixed (byte *pMetadata = metadata, pEventNameBytes = eventNameBytes)
+                {
+                    uint offset = 0;
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, eventID);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, pEventNameBytes, (uint)eventNameBytes.Length);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, (byte)0);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, keywords);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, eventVersion);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, level);
+                    foreach (var parameter in m_eventData[i].Parameters)
+                    {
+                        // Write parameter type.
+                        WriteToBuffer(pMetadata, metadataLength, ref offset, (uint)Type.GetTypeCode(parameter.ParameterType));
+                        
+                        // Write parameter name.
+                        string parameterName = parameter.Name;
+                        byte[] parameterNameBytes = Encoding.ASCII.GetBytes(parameterName);
+                        fixed (byte *pParameterNameBytes = parameterNameBytes)
+                        {
+                            WriteToBuffer(pMetadata, metadataLength, ref offset, pParameterNameBytes, (uint)parameterNameBytes.Length);
+                        }
+
+                        // Write terminated 0.
+                        WriteToBuffer(pMetadata, metadataLength, ref offset, (byte)0);
+                    }
+
+                    IntPtr eventHandle = m_provider.m_eventProvider.DefineEventHandle(eventID, eventName, keywords, eventVersion, level, pMetadata, metadataLength);
```

Comment:
Add an assert that the expected offset is now exactly at the length of the buffer.